packages/liveblocks-react/src/factory.tsx

```diff
@@ -238,22 +235,64 @@ export function createRoomContext<
   TRoomEvent,
   TThreadMetadata
 > {
-  const RoomContext = React.createContext<Room<
-    TPresence,
-    TStorage,
-    TUserMeta,
-    TRoomEvent
-  > | null>(null);
+  type TRoom = Room<TPresence, TStorage, TUserMeta, TRoomEvent>;
+
+  const RoomContext = React.createContext<TRoom | null>(null);
+
+  const roomCache = new Map<string, { room: TRoom; leave: () => void }>();
+
+  /**
+   * A cached version of .enterRoom(), where we cache each room/leave pair on
+   * a per-RoomProvider instance basis.
+   */
+  function stableEnterRoom(
+    instanceId: string,
+    roomId: string,
+    options: EnterOptions<TPresence, TStorage>
+  ): { room: TRoom; leave: () => void } {
+    const key = `${instanceId}:${roomId}`;
+    const cached = roomCache.get(key);
+    if (cached) return cached;
+
+    const rv = client.enterRoom<TPresence, TStorage, TUserMeta, TRoomEvent>(
+      roomId,
+      options
+    );
 
-  function RoomProvider(props: RoomProviderProps<TPresence, TStorage>) {
+    // Wrap the leave function to also delete the cached value
+    const origLeave = rv.leave;
+    rv.leave = () => {
+      origLeave();
+      roomCache.delete(key);
+    };
+
+    roomCache.set(key, rv);
+    return rv;
+  }
+
+  function RoomProviderOuter(props: RoomProviderProps<TPresence, TStorage>) {
+    // This "Outer" hack mostly exists to make ID generation work well in
+    // StrictMode. See this issue https://github.com/facebook/react/issues/27103#issuecomment-1763359077
+    const instanceId = React.useRef(`p${++lastInstanceId}`).current; // Like useId(), but will also work in React<18
```

Comment:
This is maybe a nitpick and I know we do this in other places (e.g. [useInitial](https://github.com/liveblocks/liveblocks/blob/main/packages/liveblocks-react/src/lib/use-initial.ts) hook)  too but React discourages reading the `ref.current` during rendering. 

https://react.dev/reference/react/useRef#caveats

> Do not write or read ref.current during rendering, except for [initialization.](https://react.dev/reference/react/useRef#avoiding-recreating-the-ref-contents) This makes your component’s behavior unpredictable.

Reply:
Wow, this is new information to me! I don't understand why that would be discouraged, but thanks for bringing this to my attention! I'll dig a bit deeper tomorrow, and see what the recommended approach would be.

Reply:
I believe the recommended way is to use `state` if something needs to be read or written to during rendering. There is slightly more information about this in the `Pitfall` section under [useRef documentation](https://react.dev/reference/react/useRef).